File: benchmarks/real_comparison/qsort.js
// QuickSort benchmark - Node.js
const SIZE = 1000000;
let arr = new Int32Array(SIZE);

function quicksort(a, low, high) {
  if (low < high) {
    const pivot = a[high];
    let i = low - 1;

    for (let j = low; j < high; j++) {
      if (a[j] <= pivot) {
        i++;
        [a[i], a[j]] = [a[j], a[i]];
      }
    }
    [a[i + 1], a[high]] = [a[high], a[i + 1]];

    const pi = i + 1;
    quicksort(a, low, pi - 1);
    quicksort(a, pi + 1, high);
  }
}

// Simple LCG for reproducible random
function lcg(seed) {
  return (seed * 1103515245 + 12345) >>> 0;
}

const start = performance.now();

for (let run = 0; run < 5; run++) {
  let seed = 42 + run;
  for (let i = 0; i < SIZE; i++) {
    seed = lcg(seed);
    arr[i] = (seed >> 16) % 1000000;
  }
  quicksort(arr, 0, SIZE - 1);
}

const end = performance.now();
console.log(
  `Node.js: QuickSort ${SIZE} elements, Time: ${(end - start).toFixed(
    2
  )} ms (5 runs)`
);

File: benchmarks/real_comparison/qsort.rs
// QuickSort benchmark - Rust
use std::time::Instant;

const SIZE: usize = 1000000;

fn quicksort(a: &mut [i32], low: i32, high: i32) {
    if low < high {
        let pivot = a[high as usize];
        let mut i = low - 1;

        for j in low..high {
            if a[j as usize] <= pivot {
                i += 1;
                a.swap(i as usize, j as usize);
            }
        }
        a.swap((i + 1) as usize, high as usize);

        let pi = i + 1;
        quicksort(a, low, pi - 1);
        quicksort(a, pi + 1, high);
    }
}

fn main() {
    let mut arr = vec![0i32; SIZE];

    let start = Instant::now();

    for run in 0..5 {
        // Simple LCG random number generator
        let mut seed: u64 = 42 + run;
        for i in 0..SIZE {
            seed = seed.wrapping_mul(1103515245).wrapping_add(12345);
            arr[i] = ((seed >> 16) % 1000000) as i32;
        }
        quicksort(&mut arr, 0, (SIZE - 1) as i32);
    }

    let elapsed = start.elapsed();
    println!(
        "Rust: QuickSort {} elements, Time: {:.2} ms (5 runs)",
        SIZE,
        elapsed.as_secs_f64() * 1000.0
    );
}
